File: movie_emoji.cs
// movie_emoji.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Threading;
using System.Diagnostics;

class MovieEmojiGame
{
    static string Colorize(string text, string color)
    {
        string col = color switch
        {
            "green" => "\x1b[92m",
            "red" => "\x1b[91m",
            "yellow" => "\x1b[93m",
            "blue" => "\x1b[94m",
            "cyan" => "\x1b[96m",
            "bold" => "\x1b[1m",
            _ => "\x1b[0m"
        };
        return col + text + "\x1b[0m";
    }

    class Movie
    {
        public string Emoji { get; set; }
        public string Title { get; set; }
        public string Hint { get; set; }
    }

    static Dictionary<string, List<Movie>> MOVIES = new Dictionary<string, List<Movie>>
    {
        {"easy", new List<Movie>{
            new Movie{Emoji="🦁👑", Title="Король Лев", Hint="Мультфильм Диснея"},
            new Movie{Emoji="🚀🌌", Title="Звёздные войны", Hint="Космическая сага"},
            new Movie{Emoji="🧙‍♂️💍", Title="Властелин колец", Hint="Фэнтези-трилогия"},
            new Movie{Emoji="🦇🃏", Title="Тёмный рыцарь", Hint="Фильм о Бэтмене"},
            new Movie{Emoji="🤖🔫", Title="Терминатор", Hint="Культовый фантастический боевик"},
            new Movie{Emoji="🦈🌊", Title="Челюсти", Hint="Фильм ужасов о большой белой акуле"},
            new Movie{Emoji="👽👾", Title="Инопланетянин", Hint="Стивен Спилберг"},
            new Movie{Emoji="🚢💖", Title="Титаник", Hint="Романтическая драма"},
            new Movie{Emoji="🐉⚔️", Title="Гарри Поттер", Hint="Фильм о волшебнике"},
            new Movie{Emoji="🧟‍♂️🧠", Title="Зомбиленд", Hint="Кинокомедия про зомби"},
        }},
        {"medium", new List<Movie>{
            new Movie{Emoji="🔫🐟", Title="Крепкий орешек", Hint="Боевик с Брюсом Уиллисом"},
            new Movie{Emoji="👻🏠", Title="Оно", Hint="Фильм ужасов о клоуне"},
            new Movie{Emoji="🧜‍♀️🧽", Title="Губка Боб", Hint="Мультфильм о морских обитателях"},
            new Movie{Emoji="🐺📈", Title="Волк с Уолл-стрит", Hint="Драма о финансисте"},
            new Movie{Emoji="🌪️🏠", Title="Унесённые ветром", Hint="Классическая мелодрама"},
            new Movie{Emoji="🐧🎬", Title="Мадагаскар", Hint="Мультфильм о животных"},
            new Movie{Emoji="🤡🎈", Title="Оно", Hint="Фильм ужасов"},
            new Movie{Emoji="🧛‍♂️🌙", Title="Дракула", Hint="Готический фильм ужасов"},
            new Movie{Emoji="🚗💨", Title="Форсаж", Hint="Скоростной боевик"},
            new Movie{Emoji="🔪😱", Title="Пила", Hint="Фильм ужасов"},
        }},
        {"hard", new List<Movie>{
            new Movie{Emoji="🦑🐟", Title="В поисках Немо", Hint="Мультфильм о рыбках"},
            new Movie{Emoji="🐭🍝", Title="Рататуй", Hint="Мультфильм о крысе-поваре"},
            new Movie{Emoji="🎪🎭", Title="Безумный Макс", Hint="Постапокалиптический боевик"},
            new Movie{Emoji="🚲🎬", Title="Велосипедист", Hint="Драма о гонщике"},
            new Movie{Emoji="🐺📖", Title="Американский психопат", Hint="Психологический триллер"},
            new Movie{Emoji="👨‍🦳🧠", Title="Игры разума", Hint="Драма о гениальном математике"},
            new Movie{Emoji="🕵️‍♂️🔍", Title="Шерлок Холмс", Hint="Детективный фильм"},
            new Movie{Emoji="🧛‍♂️❤️", Title="Сумерки", Hint="Фильм о вампирах"},
            new Movie{Emoji="🦍🏢", Title="Кинг-Конг", Hint="Приключенческий фильм"},
            new Movie{Emoji="👽🌍", Title="День независимости", Hint="Фантастический фильм"},
        }}
    };

    class Stats
    {
        public int games { get; set; }
        public int best_score { get; set; }
        public int total_movies { get; set; }
    }

    private string level;
    private List<Movie> movies;
    private int maxMovies;
    private int score;
    private int correct;
    private int skipped;
    private int hintUsed;
    private int timeLimit;
    private Stats stats;
    private string statsFile;

    public MovieEmojiGame(string lvl)
    {
        level = lvl;
        movies = MOVIES[level];
        maxMovies = 10;
        score = 0;
        correct = 0;
        skipped = 0;
        hintUsed = 0;
        statsFile = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".movie_emoji_stats.json");
        LoadStats();
        timeLimit = level == "easy" ? 60 : level == "medium" ? 45 : 30;
    }

    void LoadStats()
    {
        if (File.Exists(statsFile))
        {
            try
            {
                string json = File.ReadAllText(statsFile);
                stats = JsonSerializer.Deserialize<Stats>(json);
            }
            catch { stats = new Stats(); }
        }
        else stats = new Stats();
    }

    void SaveStats()
    {
        string json = JsonSerializer.Serialize(stats);
        File.WriteAllText(statsFile, json);
    }

    string TimerInput(string prompt, int timeout)
    {
        Console.Write(Colorize(prompt, "bold"));
        var sw = Stopwatch.StartNew();
        string input = "";
        while (sw.Elapsed.TotalSeconds < timeout)
        {
            if (Console.KeyAvailable)
            {
                input = Console.ReadLine();
                return input;
            }
            int remaining = timeout - (int)sw.Elapsed.TotalSeconds;
            Console.Write($"\r{Colorize($"Осталось времени: {remaining} сек", "yellow")}");
            Thread.Sleep(1000);
        }
        return "";
    }

    void ShowHint(Movie movie)
    {
        Console.WriteLine(Colorize($"Подсказка: первая буква '{movie.Title[0]}', всего {movie.Title.Length} букв", "green"));
    }

    void PlayRound(Movie movie)
    {
        Console.WriteLine(Colorize($"\n🎬 Эмодзи: {movie.Emoji}", "cyan"));
        Console.WriteLine(Colorize($"Подсказка: {movie.Hint}", "blue"));
        Console.WriteLine("Введите название фильма, ? для подсказки, pass для пропуска, quit для выхода.");

        var sw = Stopwatch.StartNew();
        string answer = TimerInput("Ваш ответ: ", timeLimit);
        sw.Stop();
        double elapsed = sw.Elapsed.TotalSeconds;

        if (string.IsNullOrEmpty(answer))
        {
            Console.WriteLine(Colorize("⏰ Время вышло!", "red"));
            skipped++;
            Console.WriteLine(Colorize($"Загаданный фильм: {movie.Title}", "yellow"));
            return;
        }
        if (answer == "quit")
        {
            Console.WriteLine("Выход.");
            SaveStats();
            Environment.Exit(0);
        }
        if (answer == "?")
        {
            if (hintUsed < 1)
            {
                ShowHint(movie);
                hintUsed++;
                score = Math.Max(0, score - 5);
                PlayRound(movie);
                return;
            }
            else
            {
                Console.WriteLine(Colorize("Подсказка уже использована.", "yellow"));
                PlayRound(movie);
                return;
            }
        }
        if (answer == "pass")
        {
            skipped++;
            Console.WriteLine(Colorize($"Загаданный фильм: {movie.Title}", "yellow"));
            return;
        }
        if (answer.Equals(movie.Title, StringComparison.OrdinalIgnoreCase))
        {
            correct++;
            int points = Math.Max(1, (int)(10 - elapsed / 10));
            score += points;
            Console.WriteLine(Colorize($"✅ Верно! +{points} очков. Время: {elapsed:F1} сек", "green"));
        }
        else
        {
            skipped++;
            Console.WriteLine(Colorize($"❌ Неверно. Загаданный фильм: {movie.Title}", "red"));
        }
    }

    public void Play()
    {
        Console.WriteLine(Colorize("🎬 Добро пожаловать в игру 'Угадай фильм по эмодзи'!", "bold"));
        Console.WriteLine($"Уровень: {level}, лимит времени: {timeLimit} сек.");
        Console.WriteLine("Цель: угадать фильм по набору эмодзи.");
        Console.WriteLine("Вводите название, используйте ? для подсказки, pass для пропуска, quit для выхода.\n");

        Random rnd = new Random();
        movies = movies.OrderBy(x => rnd.Next()).ToList();
        int total = Math.Min(maxMovies, movies.Count);
        for (int i = 0; i < total; i++)
        {
            Console.WriteLine(Colorize($"\nФильм {i+1}/{total}", "blue"));
            PlayRound(movies[i]);
        }
        Console.WriteLine(Colorize("\n🏁 Игра завершена!", "bold"));
        Console.WriteLine($"  Угадано фильмов: {correct}");
        Console.WriteLine($"  Пропущено: {skipped}");
        Console.WriteLine($"  Использовано подсказок: {hintUsed}");
        Console.WriteLine($"  Счёт: {score}");
        stats.games++;
        if (score > stats.best_score)
        {
            stats.best_score = score;
            Console.WriteLine(Colorize("🏆 Новый рекорд!", "green"));
        }
        stats.total_movies += correct;
        SaveStats();
        Console.WriteLine(Colorize($"Лучший результат: {stats.best_score}", "yellow"));
    }

    static void Main(string[] args)
    {
        string level = "easy";
        bool showStats = false, resetStats = false;
        foreach (var arg in args)
        {
            if (arg == "easy" || arg == "medium" || arg == "hard") level = arg;
            else if (arg == "-s" || arg == "--stats") showStats = true;
            else if (arg == "-r" || arg == "--reset") resetStats = true;
            else if (arg == "-h" || arg == "--help")
            {
                Console.WriteLine("Usage: movie_emoji [easy|medium|hard] [-s] [-r]");
                return;
            }
        }
        if (resetStats)
        {
            string f = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".movie_emoji_stats.json");
            if (File.Exists(f)) File.Delete(f);
            Console.WriteLine("Статистика сброшена.");
            return;
        }
        if (showStats)
        {
            string f = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".movie_emoji_stats.json");
            if (File.Exists(f))
            {
                try
                {
                    string json = File.ReadAllText(f);
                    var stats = JsonSerializer.Deserialize<Stats>(json);
                    Console.WriteLine(Colorize("📊 Статистика:", "bold"));
                    Console.WriteLine($"  Сыграно игр: {stats.games}");
                    Console.WriteLine($"  Лучший счёт: {stats.best_score}");
                    Console.WriteLine($"  Всего угадано фильмов: {stats.total_movies}");
                }
                catch { Console.WriteLine("Статистика пуста."); }
            }
            else Console.WriteLine("Статистика пуста.");
            return;
        }
        MovieEmojiGame game = new MovieEmojiGame(level);
        game.Play();
    }
}

File: movie_emoji.js
// movie_emoji.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const path = require('path');
const os = require('os');
const readline = require('readline');

const COLORS = {
    reset: '\x1b[0m',
    green: '\x1b[92m',
    red: '\x1b[91m',
    yellow: '\x1b[93m',
    blue: '\x1b[94m',
    cyan: '\x1b[96m',
    bold: '\x1b[1m'
};

function colorize(text, color) {
    return COLORS[color] + text + COLORS.reset;
}

const MOVIES = {
    easy: [
        { emoji: '🦁👑', title: 'Король Лев', hint: 'Мультфильм Диснея' },
        { emoji: '🚀🌌', title: 'Звёздные войны', hint: 'Космическая сага' },
        { emoji: '🧙‍♂️💍', title: 'Властелин колец', hint: 'Фэнтези-трилогия' },
        { emoji: '🦇🃏', title: 'Тёмный рыцарь', hint: 'Фильм о Бэтмене' },
        { emoji: '🤖🔫', title: 'Терминатор', hint: 'Культовый фантастический боевик' },
        { emoji: '🦈🌊', title: 'Челюсти', hint: 'Фильм ужасов о большой белой акуле' },
        { emoji: '👽👾', title: 'Инопланетянин', hint: 'Стивен Спилберг' },
        { emoji: '🚢💖', title: 'Титаник', hint: 'Романтическая драма' },
        { emoji: '🐉⚔️', title: 'Гарри Поттер', hint: 'Фильм о волшебнике' },
        { emoji: '🧟‍♂️🧠', title: 'Зомбиленд', hint: 'Кинокомедия про зомби' },
    ],
    medium: [
        { emoji: '🔫🐟', title: 'Крепкий орешек', hint: 'Боевик с Брюсом Уиллисом' },
        { emoji: '👻🏠', title: 'Оно', hint: 'Фильм ужасов о клоуне' },
        { emoji: '🧜‍♀️🧽', title: 'Губка Боб', hint: 'Мультфильм о морских обитателях' },
        { emoji: '🐺📈', title: 'Волк с Уолл-стрит', hint: 'Драма о финансисте' },
        { emoji: '🌪️🏠', title: 'Унесённые ветром', hint: 'Классическая мелодрама' },
        { emoji: '🐧🎬', title: 'Мадагаскар', hint: 'Мультфильм о животных' },
        { emoji: '🤡🎈', title: 'Оно', hint: 'Фильм ужасов' },
        { emoji: '🧛‍♂️🌙', title: 'Дракула', hint: 'Готический фильм ужасов' },
        { emoji: '🚗💨', title: 'Форсаж', hint: 'Скоростной боевик' },
        { emoji: '🔪😱', title: 'Пила', hint: 'Фильм ужасов' },
    ],
    hard: [
        { emoji: '🦑🐟', title: 'В поисках Немо', hint: 'Мультфильм о рыбках' },
        { emoji: '🐭🍝', title: 'Рататуй', hint: 'Мультфильм о крысе-поваре' },
        { emoji: '🎪🎭', title: 'Безумный Макс', hint: 'Постапокалиптический боевик' },
        { emoji: '🚲🎬', title: 'Велосипедист', hint: 'Драма о гонщике' },
        { emoji: '🐺📖', title: 'Американский психопат', hint: 'Психологический триллер' },
        { emoji: '👨‍🦳🧠', title: 'Игры разума', hint: 'Драма о гениальном математике' },
        { emoji: '🕵️‍♂️🔍', title: 'Шерлок Холмс', hint: 'Детективный фильм' },
        { emoji: '🧛‍♂️❤️', title: 'Сумерки', hint: 'Фильм о вампирах' },
        { emoji: '🦍🏢', title: 'Кинг-Конг', hint: 'Приключенческий фильм' },
        { emoji: '👽🌍', title: 'День независимости', hint: 'Фантастический фильм' },
    ]
};

class MovieEmojiGame {
    constructor(level = 'easy') {
        this.level = level;
        this.movies = MOVIES[level];
        this.maxMovies = 10;
        this.score = 0;
        this.correct = 0;
        this.skipped = 0;
        this.hintUsed = 0;
        this.statsFile = path.join(os.homedir(), '.movie_emoji_stats.json');
        this.loadStats();
        this.timeLimit = { easy: 60, medium: 45, hard: 30 }[level];
    }

    loadStats() {
        try {
            this.stats = JSON.parse(fs.readFileSync(this.statsFile, 'utf8'));
        } catch {
            this.stats = { games: 0, best_score: 0, total_movies: 0 };
        }
    }

    saveStats() {
        fs.writeFileSync(this.statsFile, JSON.stringify(this.stats, null, 2));
    }

    timerInput(prompt, timeout) {
        return new Promise((resolve) => {
            const rl = readline.createInterface({
                input: process.stdin,
                output: process.stdout
            });
            let answered = false;
            const timer = setTimeout(() => {
                if (!answered) {
                    rl.close();
                    resolve('');
                }
            }, timeout * 1000);
            rl.question(colorize(prompt, 'bold'), (answer) => {
                answered = true;
                clearTimeout(timer);
                rl.close();
                resolve(answer.trim());
            });
        });
    }

    showHint(movie) {
        console.log(colorize(`Подсказка: первая буква '${movie.title[0]}', всего ${movie.title.length} букв`, 'green'));
    }

    async playRound(movie) {
        console.log(colorize(`\n🎬 Эмодзи: ${movie.emoji}`, 'cyan'));
        console.log(colorize(`Подсказка: ${movie.hint}`, 'blue'));
        console.log('Введите название фильма, ? для подсказки, pass для пропуска, quit для выхода.');

        const start = Date.now();
        const answer = await this.timerInput('Ваш ответ: ', this.timeLimit);
        const elapsed = (Date.now() - start) / 1000;

        if (answer === '') {
            console.log(colorize('⏰ Время вышло!', 'red'));
            this.skipped++;
            console.log(colorize(`Загаданный фильм: ${movie.title}`, 'yellow'));
            return;
        }
        if (answer === 'quit') {
            console.log('Выход.');
            this.saveStats();
            process.exit(0);
        }
        if (answer === '?') {
            if (this.hintUsed < 1) {
                this.showHint(movie);
                this.hintUsed++;
                this.score = Math.max(0, this.score - 5);
                await this.playRound(movie);
                return;
            } else {
                console.log(colorize('Подсказка уже использована.', 'yellow'));
                await this.playRound(movie);
                return;
            }
        }
        if (answer === 'pass') {
            this.skipped++;
            console.log(colorize(`Загаданный фильм: ${movie.title}`, 'yellow'));
            return;
        }
        if (answer.toLowerCase() === movie.title.toLowerCase()) {
            this.correct++;
            const points = Math.max(1, Math.floor(10 - elapsed / 10));
            this.score += points;
            console.log(colorize(`✅ Верно! +${points} очков. Время: ${elapsed.toFixed(1)} сек`, 'green'));
        } else {
            this.skipped++;
            console.log(colorize(`❌ Неверно. Загаданный фильм: ${movie.title}`, 'red'));
        }
    }

    async play() {
        console.log(colorize('🎬 Добро пожаловать в игру "Угадай фильм по эмодзи"!', 'bold'));
        console.log(`Уровень: ${this.level}, лимит времени: ${this.timeLimit} сек.`);
        console.log('Цель: угадать фильм по набору эмодзи.');
        console.log('Вводите название, используйте ? для подсказки, pass для пропуска, quit для выхода.\n');

        // Перемешиваем
        for (let i = this.movies.length - 1; i > 0; i--) {
            const j = Math.floor(Math.random() * (i + 1));
            [this.movies[i], this.movies[j]] = [this.movies[j], this.movies[i]];
        }
        const total = Math.min(this.maxMovies, this.movies.length);
        for (let i = 0; i < total; i++) {
            console.log(colorize(`\nФильм ${i+1}/${total}`, 'blue'));
            await this.playRound(this.movies[i]);
        }
        console.log(colorize('\n🏁 Игра завершена!', 'bold'));
        console.log(`  Угадано фильмов: ${this.correct}`);
        console.log(`  Пропущено: ${this.skipped}`);
        console.log(`  Использовано подсказок: ${this.hintUsed}`);
        console.log(`  Счёт: ${this.score}`);
        this.stats.games++;
        if (this.score > this.stats.best_score) {
            this.stats.best_score = this.score;
            console.log(colorize('🏆 Новый рекорд!', 'green'));
        }
        this.stats.total_movies += this.correct;
        this.saveStats();
        console.log(colorize(`Лучший результат: ${this.stats.best_score}`, 'yellow'));
    }
}

async function main() {
    let level = 'easy';
    let showStats = false;
    let resetStats = false;
    const args = process.argv.slice(2);
    for (const arg of args) {
        if (arg === 'easy' || arg === 'medium' || arg === 'hard') level = arg;
        else if (arg === '-s' || arg === '--stats') showStats = true;
        else if (arg === '-r' || arg === '--reset') resetStats = true;
        else if (arg === '-h' || arg === '--help') {
            console.log('Usage: node movie_emoji.js [easy|medium|hard] [-s] [-r]');
            process.exit(0);
        }
    }
    if (resetStats) {
        const f = path.join(os.homedir(), '.movie_emoji_stats.json');
        if (fs.existsSync(f)) fs.unlinkSync(f);
        console.log('Статистика сброшена.');
        return;
    }
    if (showStats) {
        const f = path.join(os.homedir(), '.movie_emoji_stats.json');
        try {
            const stats = JSON.parse(fs.readFileSync(f, 'utf8'));
            console.log(colorize('📊 Статистика:', 'bold'));
            console.log(`  Сыграно игр: ${stats.games}`);
            console.log(`  Лучший счёт: ${stats.best_score}`);
            console.log(`  Всего угадано фильмов: ${stats.total_movies}`);
        } catch {
            console.log('Статистика пуста.');
        }
        return;
    }
    const game = new MovieEmojiGame(level);
    await game.play();
}

main().catch(console.error);
